flight/mcu/lib/FifoBuffer/FifoBuffer.h

```diff
@@ -1,67 +1,106 @@
-//
-// Created by 16182 on 9/27/2021.
-//
-
 #ifndef TARS_SOFTWARE_FIFO_H
 #define TARS_SOFTWARE_FIFO_H
 
 #include <cstdint>
 #include <cstring>
+#include <ChRt.h>
 
-#ifdef COMPILE_LOCAL
-#include <mutex>
-#endif
+template <typename T, size_t max_size>
+class FifoBuffer {
+public:
+    struct FifoIterator {
+    public:
+        using iterator_category = std::input_iterator_tag;
+        using difference_type = int64_t;
+        using value_type = T;
+        using pointer_type = T*;
+        using reference_type = T&;
 
-#ifdef COMPILE_TARGET
-#include <ChRt.h>
-#endif
-
-// only works with types that can be copied with memcpy
-class GenericFifoBuffer {
-   public:
-    // buff_size = size of array in elements
-    // data_size = size of array element in bytes
-    GenericFifoBuffer(void* arr, uint16_t buff_size, uint16_t data_size) {
-        capacity_ = buff_size;
-        cur_length_ = 0;
-        head_idx_ = 0;
-        tail_idx_ = 0;
-        data_size_ = data_size;
-        arr_ = arr;
-    }
+        FifoIterator(T* buf, size_t idx) : buf(buf), idx(idx) {}
 
-    bool push(void* element);
+        T& operator*() const {
+            return buf[idx];
+        }
+        T* operator->() {
+            return &buf[idx];
+        }
+        FifoIterator& operator++() {
+            idx = (idx + 1) % max_size;
+            return *this;
+        }
+        FifoIterator operator++(int) {
+            FifoIterator tmp = *this;
+            ++(*this);
+            return tmp;
+        }
+        friend bool operator== (const FifoIterator& a, const FifoIterator& b) {
+            return a.buf == b.buf && a.idx == b.idx;
+        }
+        friend bool operator!= (const FifoIterator& a, const FifoIterator& b) {
+            return a.buf != b.buf || a.idx != b.idx;
+        }
 
-    // returns false on failure
-    bool pop(void* out);
+    private:
+        T* buf;
+        size_t idx;
+    };
 
-   private:
-    uint16_t capacity_, cur_length_, head_idx_, tail_idx_, data_size_;
-    void* arr_;
 
-#ifdef COMPILE_LOCAL
-    std::mutex lock_;
-#endif
+    FifoBuffer() : length(0), head_idx(0), tail_idx(0) {}
 
-#ifdef COMPILE_TARGET
-    mutex_t lock_;
-#endif
-};
+    bool push(T element) {
+        lock();
+        if (length == max_size) {
+            // if length == max_size, then head_idx == tail_idx
+            memmove(&arr[tail_idx], &element, sizeof(T));
```

Comment:
why use memmove here? It was needed when the buffer was type erased, but with a monomorphized buffer its not an issue to just say arr[tail_idx] = element

Reply:
I was thinking that we might run into weird C++ copy/move semantics if we directly used the assign operator, so I used memmove to enforce shallow copy semantics.

Reply:
Using memmove will cause broken copy semantics. arr[tail] = std::move(element) is the best way to do it.

Reply:
Ah ok, I'll fix it